[Refactor:System] Add global submitty venv

cjreed121 opened the pull request:
### Why is this Change Important & Necessary?
Related to #10432

We currently install pip dependencies to the global system environment. As of Ubuntu 24.04, this is considered bad practice to do as it can break system dependencies causing unexpected issues. Thus, use of virtual environments should be done before migrating to Ubuntu 24.04.

### What is the New Behavior?
I have added a global venv `/usr/local/submitty/venv` for all scripts, daemons, etc to use. Given we were relying on a global environment before, there shouldn't be any risk of sharing this global venv across all scripts and daemons. In the future, we can consider splitting this up more if that seems worthwhile.

### What steps should a reviewer take to reproduce or test the bug or new feature?
Install the changes and verify correct behavior of various scripts, daemons, and other actions which call down to python scripts. Updating workers should definitely be tested (I have not done this yet).

### Automated Testing & Documentation
This definitely requires documentation updates and I have not made those changes yet.

### Other information
This is in draft but any reviews are still welcome.

### TODO:

- [ ] Add migration for associated `/etc/sudoers` updates
- [ ] Update documentation to tell sysadmins/shell users about the new virtual environment
- [ ] Consider if this deserves a sys admin action tag? If people wrote things to call any of our scripts, they should be aware of the dependency environment changing
- [x] Add to dev root bashrc to automatically put user in virtual environment

cjreed121 commented:
> How does this affect workers and autograding?

Take a look at the edits to the `.service` files. Those all get started inside the venv. The only concern I have with them so far is with calling scripts across machines with the proxy jump stuff we have.

cjreed121 commented:
Need to add migrations to handle the `/etc/sudoers` file updates

cjreed121 committed:
Support global virtual environment

NicholasCiuica committed:
Merge branch 'main' into add_global_venv

roye2 commented on .setup/apache/submitty.conf:
```diff
@@ -142,7 +142,7 @@
         AuthName "Git Access"
         AuthBasicProvider wsgi
-        WSGIAuthUserScript /usr/local/submitty/sbin/authentication.py
+        WSGIAuthUserScript /usr/local/submitty/venv/bin/python3 /usr/local/submitty/sbin/authentication.py
```
https://modwsgi.readthedocs.io/en/master/configuration-directives/WSGIPythonHome.html
https://modwsgi.readthedocs.io/en/master/configuration-directives/WSGIAuthUserScript.html
https://modwsgi.readthedocs.io/en/master/configuration-directives/WSGIDaemonProcess.html

It looks like apache setup is failing here because you can't set a python virtual environment for this particular config in WSGI. Rather, you can set the python venv with the WSGIPythonHome globally, or set it for the daemon process with WSGIDaemonProcess and then tell WSGIAuthUserScript to use it. I may make a commit to this PR to test that functionality.

cjreed121 replied:
There are changes in https://github.com/Submitty/Submitty/pull/12849 which need to be moved over here. One of those is what addresses this

NicholasCiuica committed:
Testing if this is causing issues

This is removed in the Ubuntu 24.04 upgrade PR, which is planning to push to this venv PR, so I'm thinking this is not a needed change

NicholasCiuica committed:
shell linting double quotes fix

NicholasCiuica committed:
shell linting source filename possible fix

NicholasCiuica committed:
Merge branch 'main' into add_global_venv

NicholasCiuica committed:
trying -x flag on shellcheck

NicholasCiuica committed:
Revert "trying -x flag on shellcheck"

This reverts commit a1071f0e4da34a9be58a12e4a358c603865618f9.

NicholasCiuica committed:
add shellcheck ignore to source venv/bin/activate

NicholasCiuica committed:
Merge branch 'main' into add_global_venv

NicholasCiuica committed:
moving the shellcheck ignore to the right location

Accidentally put it on the wrong source command in a different file lol

NicholasCiuica committed:
comment got read as shellcheck

NicholasCiuica committed:
ansible error; is indentation wrong?

NicholasCiuica committed:
Merge branch 'main' into add_global_venv

dagemcn commented:
<img width="1676" height="87" alt="image" src="https://github.com/user-attachments/assets/d68d9469-a469-4306-b8d7-e75b14cfe9fa" />
I encountered this error when trying to run submitty_install on this branch. It looks like this could be an arm issue as this was on Mac, though I'm not sure.

NicholasCiuica committed:
Merge branch 'main' into add_global_venv

NicholasCiuica committed:
trying to remove echo input

NicholasCiuica committed:
show errors in CI

NicholasCiuica committed:
Merge branch 'main' into add_global_venv

NicholasCiuica committed:
trying to run command as sudo

NicholasCiuica committed:
Merge branch 'add_global_venv' of https://github.com/Submitty/Submitty into add_global_venv

NicholasCiuica committed:
not sure how to run install_system as root in ansible

NicholasCiuica committed:
adding back log file

NicholasCiuica committed:
adding back previous code

NicholasCiuica committed:
Merge branch 'main' into add_global_venv

NicholasCiuica committed:
unneeded venv in install_system?

NicholasCiuica committed:
Merge branch 'add_global_venv' of https://github.com/Submitty/Submitty into add_global_venv

roye2 commented:
<img width="1090" height="1166" alt="image" src="https://github.com/user-attachments/assets/396d3caf-ada0-4c0f-8413-71e5dced5289" />

When attempting to perform a from-scratch install, I received this error.

NicholasCiuica committed:
Revert "unneeded venv in install_system?"

This reverts commit c5b4c0071f0a2f0a348078773bc033ba3d6ad9f2.

NicholasCiuica committed:
moving install_system changes over from ubuntu PR

NicholasCiuica committed:
Merge branch 'main' into add_global_venv

NicholasCiuica committed:
missing character

NicholasCiuica committed:
Merge branch 'add_global_venv' of https://github.com/Submitty/Submitty into add_global_venv

NicholasCiuica committed:
ubuntu -> venv, INSTALL_SUBMITTY_HELPER.sh

NicholasCiuica committed:
Merge branch 'main' into add_global_venv

NicholasCiuica committed:
adduser.py now uses venv as interpreter

NicholasCiuica committed:
Merge branch 'main' into add_global_venv

NicholasCiuica committed:
changing interpreter of adduser_course.py

NicholasCiuica committed:
Merge branch 'add_global_venv' of https://github.com/Submitty/Submitty into add_global_venv

NicholasCiuica committed:
added activate to install_migrator

NicholasCiuica committed:
trying to fix course issue

NicholasCiuica committed:
Merge branch 'main' into add_global_venv

NicholasCiuica committed:
linter comment fix

NicholasCiuica committed:
Revert "trying to fix course issue"

This reverts commit d0432c4ae5c27231f5e4183ef0d80fdd6e64f8f0.

NicholasCiuica committed:
Revert "changing interpreter of adduser_course.py"

This reverts commit d48d0e5d0fdc3bb2cbb3b4600796d2ab74469b26.

NicholasCiuica committed:
Revert "adduser.py now uses venv as interpreter"

This reverts commit 744e86de3c8a620a91ceb250d1b7d17c73d2d58d.

NicholasCiuica committed:
defining venv on call instead of using shebangs

NicholasCiuica committed:
adding venv to create_course

NicholasCiuica committed:
cleaning up activate

NicholasCiuica committed:
shell lint fix

NicholasCiuica committed:
ansible lint fix

NicholasCiuica committed:
Merge branch 'main' into add_global_venv

roye2 commented:
<img width="2050" height="1166" alt="image" src="https://github.com/user-attachments/assets/79ea495f-63b4-414e-9f6c-49652704d07d" />


I ran submitty_install and received this error.

Edit, as of 7/29 I am no longer receiving this error.

NicholasCiuica committed:
Merge branch 'main' into add_global_venv

NicholasCiuica reviewed (commented):
After extensive testing and debugging, there still major issues barring progress on this PR, based on issues people are having. I'll list them here:

1) There is an issue with the run_migrator.py script or the migrations that it's running that causes issues when trying to install this PR (might be platform-specific? has happened on Mac but not on my Windows machine).
2) File permissions may be getting messed up, possibly due to migrations issue from (1) failing silently on some people's machines? Not sure about this one.
3) Some part of course creation is failing and results in courses being made without any users added to them, might be related to an issue in sudo or messed up permissions as described in (2)?

NicholasCiuica committed:
testing sudoers without venv

roye2 committed:
fix typos in create_course.sh & set_config_permissions.py

NicholasCiuica committed:
Merge branch 'main' into add_global_venv

roye2 committed:
Revert "testing sudoers without venv"

This reverts commit b6597ff7f93a8d5164c0ca0a497753930d92d3e5.

roye2 committed:
Merge remote-tracking branch 'origin/fix-course-creation' into add_global_venv

roye2 committed:
add * wildcard to course creation sudoers
this will make sure the scripts accept input arguments with sudo

roye2 committed:
Revert "Merge remote-tracking branch 'origin/fix-course-creation' into add_global_venv"

This reverts commit dbe390c96d0039795bac11326cf89ca82187ff6a, reversing
changes made to a329cd22992b81b458a46f0699daccda927f8191.

roye2 commented:
Course creation should now be working on this PR. Not sure about the migration issues. Also, I have trouble running install_system.sh locally.

roye2 committed:
update build_homework_function & install process
This should address the issue with restart_shipper_and_all_workers not
restarting things properly (missing package since it wasn't executing
with venv)

roye2 committed:
potential fix to migrator issue?

roye2 committed:
Merge branch 'main' into add_global_venv

roye2 committed:
add venv to dev root bashrc

roye2 committed:
Merge branch 'main' into add_global_venv

roye2 commented:
Added the venv to the bashrc for the Dev VM so that when you log in you're automatically put in the venv. We could consider adding a message telling developers how to deactivate. It is literally `(venv) root@vagrant:/usr/local/submitty# deactivate`

NicholasCiuica committed:
Merge branch 'main' into add_global_venv

NicholasCiuica commented:
> <img alt="image" width="2000" height="1138" src="https://private-user-images.githubusercontent.com/229643694/628000391-79ea495f-63b4-414e-9f6c-49652704d07d.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3ODYwMjQzNTEsIm5iZiI6MTc4NjAyNDA1MSwicGF0aCI6Ii8yMjk2NDM2OTQvNjI4MDAwMzkxLTc5ZWE0OTVmLTYzYjQtNDE0ZS05ZjZjLTQ5NjUyNzA0ZDA3ZC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjYwODA2JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI2MDgwNlQxMzQ3MzFaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0wMWE4ZjIzZDBmNTU3MGQxZmZmZGZhYjE5NDViMWYwZDRlMTE5NjMxOWVhNTU0YTViOTcyNzk0NWYzYWZiYTUyJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZyZXNwb25zZS1jb250ZW50LXR5cGU9aW1hZ2UlMkZwbmcifQ.1ej00GiEWcneF4s9B_LmX-iucy13QeVzPOh0eoPDpqg">

Emma thought that something may be going wrong in INSTALL_SUBMITTY_HELPER.sh that is messing with the submitty_utils file permissions, so I ran the following in the VM to double check it (a snippet like this can be found around line 180 in that file):

rsync -rtz "/usr/local/submitty/GIT_CHECKOUT/Submitty/python_submitty_utils" "usr/submitty/install"
cd "usr/submitty/install/python_submitty_utils"
(umask 022 && pip3 install .)

and then 

ls -la /usr/local/submitty/venv/lib/python3.10/site-packages/submitty_utils

<img width="944" height="496" alt="image" src="https://github.com/user-attachments/assets/7cf435c4-b0f1-4686-8a75-7783828965b7" />

The contents of submitty_utils in the venv should look something like this. However, my permissions for all files in this folder are rw-r----- (missing read permission for global), which we thought was causing the permissions issue. However, when I used chmod to manually set the file permissions in this folder to rw-r--r--, I still got the same permissions issue.

After more debugging around the "pip3 install" step (trying --no-cache-dir and --force-reinstall flags, comparing logs generated by -v flag), we still haven't found what might be causing this permissions issue on my end. We're also not sure how Emma got this issue, or how it resolved itself on her end.

NicholasCiuica commented:
We're still not sure why pip sometimes changes the permissions of the submitty_utils dependency when its installed into the venv, but Emma implemented a fix that lets us to ignore that issue for now and debug other things. See lines 188 and 189 of INSTALL_SUBMITTY_HELPER.sh in the image below.

<img width="659" height="334" alt="image" src="https://github.com/user-attachments/assets/9cb1ee78-59d5-4b10-a8cc-60aa7a2453c3" />
